File: Sum of Left Leaves/sum.cpp
/*求所有左子叶的和，运用到的二叉树的遍历，递归，添加一个helper方法用来表示当前节点是不是左子叶*/
/*先序遍历*/

class Solution {
public:
  int sumOfLeftLeaves(TreeNode* root) {
    if (!root || (!root->left && !root->right)) {
      return 0;
    }
    int res = 0;
    helper(root->left, true, res);
    helper(root->right, false, res);
    return res;
  }

  void helper(TreeNode* node, bool left, int& res) {
    if (!node) {
      return;
    }
    if (!node->left && !node->right && left)
      res += node->val;
    helper(node->left, true, res);
    helper(node->right, false, res);
  }
};

File: Sum of Left Leaves/sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sumOfLeftLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def sumLeftLeaves(node, part=0, left=0):
            if node is None:
                return 0
            if node.left or node.right:
                # Branch
                return sumLeftLeaves(node.left, part, 1) + sumLeftLeaves(node.right, part, 0)
            # Leaf
            return node.val if left else 0
        return sumLeftLeaves(root)
